sklearn/metrics/scorer.py

```diff
@@ -126,7 +126,13 @@ def __call__(self, clf, X, y, sample_weight=None):
         y_type = type_of_target(y)
         y_pred = clf.predict_proba(X)
         if y_type == "binary":
-            y_pred = y_pred[:, 1]
+            if y_pred.shape[1] == 2:
+                y_pred = y_pred[:, 1]
+            else:
+                raise ValueError('got predict_proba of shape;'
```

Comment:
aren't you meant to output the shape here?

Reply:
Why is it a good idea to output the shape here?  Isn't this more informative:  
"vector of probabilities returnedpredict_proba is all 1's because only one classifier is in the model."

Reply:
do you mean "only one class" rather than "only one classifier"?

Either way the proposed error message here says "got predict of shape; ..." that simply doesn't make sense.

Reply:
How about:  
"vector of probabilities returned (`predict_proba`) is all 1's because there is only one class in the input to the model."

Reply:
It can also fail with GMMs where the shape would be wrong but the entries wouldn't be all ones.

Reply:
How about:  
"only one vector of probabilities is returned (predict_proba) because there is only one class or cluster in the input to the model."

Reply:
Well if you add "but two are required" then it's fine. The message should say exactly what's wrong (predict_proba has shape (n_samples, 1)), what the cause is (single class or single probability output) and how to fix it (use a binary classifier).
Using it with a GMM with two components would work, but generally doesn't make sense, so I wouldn't recommend that as a solution. Mentioning clustering might be a bit confusing in general because it's not a valid use-case of this.

Reply:
But, isn't the shape:  (n_samples, 2) ?
Is the problem that they are using the wrong classifier, or rather their data is not appropriate for this algorithm?  It wouldn't be helpful to tell them to "use a binary classifier" if their data does not have 2 or more classes. 

---
Updated text:  
predict_proba has shape (n_samples, 2) which returns only 1 vector of probabilities (because it is a single class) but 2 classes in the data are required.

Reply:
If the shape is ``(n_samples, 2)``  then everything is fine. The problem is if it's ``(n_samples, 1)`` which is when the original error happens.

Reply:
right, right, got it. 
How is this:  
>predict_proba has shape (n_samples, 1) which returns only 1 vector of probabilities (because it is a single class) but 2 classes in the data are required.

Reply:
yes, but you should use actual ``y_pred.shape[1]`` not hard-code 1 as there might be code-paths that have >2. (unless we can make sure that's not happening)

Reply:
you mean like this?
```python
            if y_pred.shape[1] == 2:
                y_pred = y_pred[:, 1]
            elif y_pred.shape[1] == 1:
                raise ValueError('got predict_proba of shape;'
```

Reply:
Or, I think you mean this:

>predict_proba has shape (n_samples, y_pred.shape[1]) which returns only 1 vector of probabilities (because it is a single class) but 2 classes in the data are required.

Reply:
No I mean like
```python
ValueError("got predict_proba of shape {}, but need classifier with two classes for {} scoring".format(y_pred.shape, self._score_func.__name__))
```